modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -3192,8 +3192,7 @@ void GDScriptAnalyzer::reduce_preload(GDScriptParser::PreloadNode *p_preload) {
 		if (!FileAccess::exists(p_preload->resolved_path)) {
 			push_error(vformat(R"(Preload file "%s" does not exist.)", p_preload->resolved_path), p_preload->path);
 		} else {
-			// TODO: Don't load if validating: use completion cache.
-			p_preload->resource = ResourceLoader::load(p_preload->resolved_path);
+			p_preload->resource = GDScriptCache::get_shallow_script(p_preload->resolved_path, parser->script_path);
```

Comment:
This is where it gets weird. There's two behaviors, depending on the code on this line.

1. With `GDScriptCache::get_shallow_script()` as is:
    - my minimal reproduction project :heavy_check_mark: ;
    - the new test cases :x: ;
    - `modules/gdscript/tests/scripts/analyzer/features/preload_constant_types_are_inferred.gd` :x: ;
    - build does not fail due to memory leaks :heavy_check_mark: .
3. With `ResourceLoader::load()` and then `GDScriptCache::get_shallow_script()` if `p_preload->resource.is_null()`:
    - my minimal reproduction project :x: ;
    - the new test cases :heavy_check_mark: ;
    - `modules/gdscript/tests/scripts/analyzer/features/preload_constant_types_are_inferred.gd` :heavy_check_mark: ;
    - build fails due to memory leaks :warning: